crates/ty_python_semantic/src/types/infer/builder.rs

```diff
@@ -2169,6 +2181,37 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
             definition,
             &DeclaredAndInferredType::are_the_same_type(inferred_ty),
         );
+
+        if function_decorators.contains(FunctionDecorators::OVERLOAD) {
+            for stmt in &function.body {
+                match stmt {
+                    ast::Stmt::Pass(_) => continue,
+                    ast::Stmt::Expr(ast::StmtExpr { value, .. }) => {
+                        if matches!(
+                            &**value,
+                            ast::Expr::StringLiteral(_) | ast::Expr::EllipsisLiteral(_)
+                        ) {
+                            continue;
+                        }
+                    }
+                    _ => {}
+                }
+                let Some(builder) = self.context.report_lint(&USELESS_OVERLOAD_BODY, stmt) else {
+                    continue;
+                };
+                let mut diagnostic = builder.into_diagnostic(format_args!(
+                    "Useless body for `@overload`-decorated function `{}`",
+                    &function.name
+                ));
+                diagnostic.set_primary_message("This statement will never be executed");
+                diagnostic.info(
+                    "`@overload`-decorated functions are solely for type checkers \
+                    and must be overridden at runtime by a non-`@overload`-decorated implementation",
```

Comment:
Nit: "overridden" doesn't feel like the right term here. That term is usually applied to subclass overrides, not to subsequent shadowing definitions. In general I don't think it's necessarily clear to the reader what "overridden at runtime" means.

I would probably just say something like "... are solely to provide a signature for type checkers; their bodies are unused"

Reply:
switched to "overwritten", which I think is more precise, and avoids the confusion about "overidden" usually being applied to subclass overrides